### PR TITLE
prov/gni: Implement FI_MULTI_RECV

### PR DESCRIPTION
**Pull request by ztiffany**
-Match several send requests to a single receive request.
-Use req->send_len consistently to hold the real number of bytes to be moved.
-Test for truncated, unaligned send buffer earlier to allow send_len to hold
actual number of bytes moved.
-Use req->send_len to fill a CQ entry's length field.
-Rename GNIX_MSG_DOUBLE_GET to GNIX_MSG_GET_TAIL for clarity.

Fixes ofi-cray/libfabric-cray#580.

Signed-off-by: Zach Tiffany ztiffany@cray.com

@hppritcha @sungeunchoi @jswaro @chuckfossen

### CODE REVIEW

File: prov/gni/include/gnix_ep.h [hunk 1]
```diff
@@ -37,6 +37,9 @@
 #include "gnix.h"
 #include "gnix_nic.h"
 
+/* Default minimum multi receive buffer size. */
+#define GNIX_OPT_MIN_MULTI_RECV_DEFAULT	64
+
```
Comment (sungeunchoi) [inline prov/gni/include/gnix_ep.h:42]: Is this something we'd want to allow users to set?

Comment (ztiffany) [inline prov/gni/include/gnix_ep.h:42]: This value is supposed to be able to be set/retrieved using fi_getopt/fi_setopt.

```
   fi_getopt / fi_setopt
       Endpoint protocol operations may be retrieved using fi_getopt or set using fi_setopt.  Applications specify the level that  a
       desired option exists, identify the option, and provide input/output buffers to get or set the option.  fi_setopt provides an
       application a way to adjust low-level protocol and implementation specific details of an endpoint.

       The following option levels and option names and parameters are defined.

       FI_OPT_ENDPOINT

       · FI_OPT_MIN_MULTI_RECV - size_t : Defines the minimum receive buffer space available when the receive  buffer  is  automati-
         cally  freed  (see  FI_MULTI_RECV).   Modifying  this  value  is  only guaranteed to set the minimum buffer space needed on
         receives posted after the value has been changed.  It is recommended that applications that want to  override  the  default
         MIN_MULTI_RECV value set this option before enabling the corresponding endpoint.
```

Comment (sungeunchoi) [inline prov/gni/include/gnix_ep.h:42]: Which we haven't implemented yet.  We should open an issue.

File: prov/gni/src/gnix_msg.c [hunk 2]
```diff
@@ -164,10 +180,11 @@ static inline int __gnix_msg_recv_completion(struct gnix_fid_ep *ep,
...
 
 	flags |= req->msg.send_flags & (FI_TAGGED | FI_REMOTE_CQ_DATA);
-	flags |= req->msg.recv_flags & (FI_PEEK | FI_CLAIM | FI_DISCARD);
+	flags |= req->msg.recv_flags & (FI_PEEK | FI_CLAIM | FI_DISCARD |
+					FI_MULTI_RECV);
 
 	return __recv_completion(ep, req, req->user_context, flags,
-				 req->msg.recv_len, (void *)req->msg.recv_addr,
+				 req->msg.send_len, (void *)req->msg.recv_addr,
 				 req->msg.imm, req->msg.tag,
```
Comment (hppritcha) [inline prov/gni/src/gnix_msg.c:188]: just curious why the change here from `recv_len` to `send_len` in the call to `__recv_completion`?

**Comment by hppritcha** [comment]
:+1:

File: prov/gni/src/gnix_msg.c (hunk 2 above)
Comment (ztiffany) [inline prov/gni/src/gnix_msg.c:188]: Previously, we would do an extra step of storing the number of moved bytes into the recv_len field right before generating these event.  Now, the request's send_len field should always hold the count of moved bytes when the request is completed.  So the send_len field is now the best place to get this information.
